native/src/main/java/io/ballerina/stdlib/mqtt/client/ClientActions.java

```diff
@@ -98,11 +195,28 @@ public static Object externReconnect(BObject clientObject) {
         return null;
     }
 
-    private static MqttMessage generateMqttMessage(BMap message) {
-        MqttMessage mqttMessage = new MqttMessage();
-        mqttMessage.setPayload(((BArray) message.get(StringUtils.fromString(MqttConstants.PAYLOAD))).getByteArray());
-        mqttMessage.setQos(((Long) message.get(StringUtils.fromString(MqttConstants.QOS))).intValue());
-        mqttMessage.setRetained(((boolean) message.get(StringUtils.fromString(MqttConstants.RETAINED))));
-        return mqttMessage;
+    public static Object nextResult(Environment env, BObject streamIterator) {
+        BlockingQueue<?> messageQueue = (BlockingQueue<?>) streamIterator.getNativeData(RESPONSE_QUEUE);
+        ExecutorService executor = (ExecutorService) streamIterator.getNativeData(RESPONSE_EXECUTOR_SERVICE);
+        Future future = env.markAsync();
+        executor.submit(() -> {
+            try {
+                BMap message = (BMap) messageQueue.take();
+                future.complete(message);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                future.complete(MqttUtils.createMqttError(e));
+            }
+        });
+        return null;
+    }
+
+    public static Object closeStream(BObject streamIterator) {
```

Comment:
Unless there is an exception we don't need to return an `Object` from this method. And from the ballerina side we can use a nillable return type for the extern function. But, in case there is an runtime-exception thrown from this method the program will panic unless you properly handle the exception (doesn't matter you have returned an Object in that case).

Reply:
Fixed in https://github.com/ballerina-platform/module-ballerina-mqtt/pull/9/commits/041d5b9016d02da09f39901fe4d4970544edfb25